File: mobile/packages/bmoni_embedded_sdk/ios/bmoni_embedded_sdk/Sources/bmoni_embedded_sdk/BmoniEmbeddedSdkPlugin.swift
import BMONISigner
import Flutter
import Foundation
import os.log

/// Wallet/signing methods exposed by the plugin's `FlutterMethodChannel`.
///
/// The raw value of each case is the string the Dart side sends via
/// `MethodChannel.invokeMethod`. Adding a new case forces the `switch`
/// dispatch in `BmoniEmbeddedSdkPlugin.handle` to handle it.
enum ChannelMethods: String, CaseIterable {
  case initializeWallet = "initWallet"
  case signTransactionHash = "signTransactionHash"
  case signMessage = "signMessage"
  case deleteWallet = "deleteWallet"
}

/// Flutter plugin that bridges the Dart facade to the BMONISigner iOS
/// framework (Swift global functions: `initWallet`, `signTransactionHash`,
/// `signMessage`, `deleteWallet`).
///
/// All signer calls are dispatched on a background queue because they
/// touch the file system and Secure Enclave; results are forwarded to
/// Flutter on the main thread.
public class BmoniEmbeddedSdkPlugin: NSObject, FlutterPlugin {
  private static let channelName = "bmoni_embedded_sdk"
  private static let signerErrorCode = "BMONI_SIGNER_ERROR"
  private static let pluginErrorCode = "BMONI_PLUGIN_ERROR"
  private static let invalidArgumentCode = "INVALID_ARGUMENT"

  private static let logger = OSLog(
    subsystem: "me.bkey.bmoni_embedded_sdk",
    category: "plugin"
  )

  private let workQueue = DispatchQueue(
    label: "me.bkey.bmoni_embedded_sdk.signer",
    qos: .userInitiated
  )

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: channelName,
      binaryMessenger: registrar.messenger()
    )
    let instance = BmoniEmbeddedSdkPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(
    _ call: FlutterMethodCall,
    result: @escaping FlutterResult
  ) {
    guard let method = ChannelMethods(rawValue: call.method) else {
      result(FlutterMethodNotImplemented)
      return
    }

    switch method {
    case .initializeWallet:
      runOnWorkQueue(result: result) {
        try initWallet()
      }

    case .signTransactionHash:
      guard
        let args = call.arguments as? [String: Any],
        let hashHex = args["hashHex"] as? String
      else {
        result(Self.missingArgument("hashHex"))
        return
      }
      runOnWorkQueue(result: result) {
        try signTransactionHash(hashHex)
      }

    case .signMessage:
      guard
        let args = call.arguments as? [String: Any],
        let message = args["message"] as? String
      else {
        result(Self.missingArgument("message"))
        return
      }
      runOnWorkQueue(result: result) {
        try signMessage(message)
      }

    case .deleteWallet:
      runOnWorkQueue(result: result) { () -> Any? in
        try deleteWallet()
        return nil
      }
    }
  }

  /// Runs [block] on the background work queue and forwards either the
  /// returned value or a typed [FlutterError] back to Flutter on the main
  /// queue.
  private func runOnWorkQueue(
    result: @escaping FlutterResult,
    block: @escaping () throws -> Any?
  ) {
    workQueue.async {
      let outcome: Outcome
      do {
        outcome = .success(try block())
      } catch let error as BMONISignerError {
        outcome = .signerFailure(error)
      } catch {
        // Surface the unexpected failure in the system log before
        // wrapping it: the consumer only sees the generic
        // `BMONI_PLUGIN_ERROR` code on the Dart side, so we want a
        // breadcrumb in Console.app / Xcode for production triage.
        os_log(
          "Unexpected error in BMONISigner bridge: %{public}@",
          log: Self.logger,
          type: .error,
          String(describing: error)
        )
        outcome = .unexpectedFailure(error)
      }

      DispatchQueue.main.async {
        switch outcome {
        case .success(let value):
          result(value)
        case .signerFailure(let error):
          result(Self.flutterError(from: error))
        case .unexpectedFailure(let error):
          result(
            FlutterError(
              code: Self.pluginErrorCode,
              message: error.localizedDescription,
              details: nil
            )
          )
        }
      }
    }
  }

  private static func missingArgument(_ name: String) -> FlutterError {
    FlutterError(
      code: invalidArgumentCode,
      message: "Missing required argument: \(name)",
      details: nil
    )
  }

  private static func flutterError(from error: BMONISignerError) -> FlutterError {
    let status: Int
    let message: String
    switch error {
    case .initError(let s, let m),
      .initWalletError(let s, let m),
      .signTransactionError(let s, let m):
      status = s
      message = m
    }

    return FlutterError(
      code: signerErrorCode,
      message: message,
      details: ["errorCode": status]
    )
  }

  private enum Outcome {
    case success(Any?)
    case signerFailure(BMONISignerError)
    case unexpectedFailure(Error)
  }
}

File: mobile/packages/bmoni_embedded_sdk/android/src/main/kotlin/com/bmoni/embedded/sdk/BmoniEmbeddedSdkPlugin.kt
package com.bmoni.embedded.sdk

import android.content.Context
import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import me.bkey.ip.bmonisigner.BMONISigner
import me.bkey.ip.bmonisigner.BMONISignerException

/**
 * Wallet/signing methods exposed by the plugin's [MethodChannel].
 *
 * The [methodName] of each entry is the string the Dart side sends via
 * `MethodChannel.invokeMethod`. Adding a new entry forces the `when`
 * dispatch in [BmoniEmbeddedSdkPlugin.onMethodCall] to handle it.
 */
enum class ChannelMethods(val methodName: String) {
    INITIALIZE_WALLET("initWallet"),
    SIGN_TRANSACTION_HASH("signTransactionHash"),
    SIGN_MESSAGE("signMessage"),
    DELETE_WALLET("deleteWallet"),
    ;

    companion object {
        private val byMethodName: Map<String, ChannelMethods> =
            entries.associateBy(ChannelMethods::methodName)

        fun fromMethodName(methodName: String): ChannelMethods? = byMethodName[methodName]
    }
}

/**
 * Flutter plugin that bridges the Dart facade to the native BMONISigner Android SDK
 * (`me.bkey.ip.bmonisigner.BMONISigner`).
 *
 * All signer calls are dispatched on a background executor because they touch the file system and
 * perform OpenSSL-backed crypto. Results are delivered back to Flutter on the main thread.
 */
class BmoniEmbeddedSdkPlugin :
    FlutterPlugin,
    MethodCallHandler {
    private companion object {
        const val CHANNEL_NAME = "bmoni_embedded_sdk"
        const val INVALID_ARGUMENT_CODE = "INVALID_ARGUMENT"
        const val SIGNER_ERROR_CODE = "BMONI_SIGNER_ERROR"
        const val PLUGIN_ERROR_CODE = "BMONI_PLUGIN_ERROR"
    }

    private lateinit var channel: MethodChannel
    private lateinit var applicationContext: Context

    private val executor: ExecutorService =
        Executors.newSingleThreadExecutor { runnable ->
            Thread(runnable, "bmoni-embedded-sdk").apply { isDaemon = true }
        }
    private val mainHandler = Handler(Looper.getMainLooper())

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, CHANNEL_NAME)
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(
        call: MethodCall,
        result: Result,
    ) {
        val method = ChannelMethods.fromMethodName(call.method)
        if (method == null) {
            result.notImplemented()
            return
        }

        when (method) {
            ChannelMethods.INITIALIZE_WALLET -> runOnExecutor(result) {
                BMONISigner.initWallet(applicationContext)
            }

            ChannelMethods.SIGN_TRANSACTION_HASH -> {
                val hashHex = call.argument<String>("hashHex")
                if (hashHex == null) {
                    result.missingArgument("hashHex")
                    return
                }
                runOnExecutor(result) {
                    BMONISigner.signTransactionHash(applicationContext, hashHex)
                }
            }

            ChannelMethods.SIGN_MESSAGE -> {
                val message = call.argument<String>("message")
                if (message == null) {
                    result.missingArgument("message")
                    return
                }
                runOnExecutor(result) {
                    BMONISigner.signMessage(applicationContext, message)
                }
            }

            ChannelMethods.DELETE_WALLET -> runOnExecutor(result) {
                BMONISigner.deleteWallet(applicationContext)
                null
            }
        }
    }

    /**
     * Executes [block] on the background [executor] and forwards either the returned value or a
     * typed error back to Flutter on the main thread.
     */
    private fun runOnExecutor(
        result: Result,
        block: () -> Any?,
    ) {
        executor.execute {
            val outcome: Outcome =
                try {
                    Outcome.Success(block())
                } catch (e: BMONISignerException) {
                    Outcome.SignerFailure(e)
                } catch (e: Exception) {
                    // Deliberately narrower than Throwable so JVM `Error`s
                    // (OutOfMemoryError, StackOverflowError, …) propagate
                    // and crash loudly instead of being swallowed and
                    // forwarded to Flutter as a recoverable error.
                    Outcome.UnexpectedFailure(e)
                }

            mainHandler.post {
                when (outcome) {
                    is Outcome.Success -> result.success(outcome.value)
                    is Outcome.SignerFailure -> {
                        val details = mapOf("errorCode" to outcome.exception.errorCode)
                        result.error(
                            SIGNER_ERROR_CODE,
                            outcome.exception.message ?: "BMONISigner error",
                            details,
                        )
                    }
                    is Outcome.UnexpectedFailure -> result.error(
                        PLUGIN_ERROR_CODE,
                        outcome.error.message ?: outcome.error.javaClass.name,
                        null,
                    )
                }
            }
        }
    }

    private fun Result.missingArgument(name: String) {
        error(INVALID_ARGUMENT_CODE, "Missing required argument: $name", null)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        executor.shutdown()
    }
}

/** Result wrapper for background work executed by [BmoniEmbeddedSdkPlugin]. */
private sealed class Outcome {
    data class Success(val value: Any?) : Outcome()
    data class SignerFailure(val exception: BMONISignerException) : Outcome()
    data class UnexpectedFailure(val error: Throwable) : Outcome()
}
